// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: envoy/extensions/certificate_providers/local_certificate/v3/local_certificate.proto

package local_certificatev3

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on LocalCertificate with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *LocalCertificate) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on LocalCertificate with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// LocalCertificateMultiError, or nil if none found.
func (m *LocalCertificate) ValidateAll() error {
	return m.validate(true)
}

func (m *LocalCertificate) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetRootcaCert()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, LocalCertificateValidationError{
					field:  "RootcaCert",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, LocalCertificateValidationError{
					field:  "RootcaCert",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetRootcaCert()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return LocalCertificateValidationError{
				field:  "RootcaCert",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if all {
		switch v := interface{}(m.GetRootcaKey()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, LocalCertificateValidationError{
					field:  "RootcaKey",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, LocalCertificateValidationError{
					field:  "RootcaKey",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetRootcaKey()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return LocalCertificateValidationError{
				field:  "RootcaKey",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if all {
		switch v := interface{}(m.GetDefaultIdentityCert()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, LocalCertificateValidationError{
					field:  "DefaultIdentityCert",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, LocalCertificateValidationError{
					field:  "DefaultIdentityCert",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetDefaultIdentityCert()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return LocalCertificateValidationError{
				field:  "DefaultIdentityCert",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if all {
		switch v := interface{}(m.GetDefaultIdentityKey()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, LocalCertificateValidationError{
					field:  "DefaultIdentityKey",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, LocalCertificateValidationError{
					field:  "DefaultIdentityKey",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetDefaultIdentityKey()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return LocalCertificateValidationError{
				field:  "DefaultIdentityKey",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if all {
		switch v := interface{}(m.GetExpirationTime()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, LocalCertificateValidationError{
					field:  "ExpirationTime",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, LocalCertificateValidationError{
					field:  "ExpirationTime",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetExpirationTime()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return LocalCertificateValidationError{
				field:  "ExpirationTime",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	// no validation rules for Pkey

	if wrapper := m.GetCacheSize(); wrapper != nil {

		if wrapper.GetValue() <= 0 {
			err := LocalCertificateValidationError{
				field:  "CacheSize",
				reason: "value must be greater than 0",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

	}

	if len(errors) > 0 {
		return LocalCertificateMultiError(errors)
	}

	return nil
}

// LocalCertificateMultiError is an error wrapping multiple validation errors
// returned by LocalCertificate.ValidateAll() if the designated constraints
// aren't met.
type LocalCertificateMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m LocalCertificateMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m LocalCertificateMultiError) AllErrors() []error { return m }

// LocalCertificateValidationError is the validation error returned by
// LocalCertificate.Validate if the designated constraints aren't met.
type LocalCertificateValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e LocalCertificateValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e LocalCertificateValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e LocalCertificateValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e LocalCertificateValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e LocalCertificateValidationError) ErrorName() string { return "LocalCertificateValidationError" }

// Error satisfies the builtin error interface
func (e LocalCertificateValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sLocalCertificate.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = LocalCertificateValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = LocalCertificateValidationError{}
